### PR TITLE
Alignment of mpi- and device-level tags.

### PR DESCRIPTION
**Pull request by alexander-sannikov**
Device level is provided with an abillity to adjust tag size on MPI-level.

`MPIR_TAG_ERROR_BIT/MPIR_TAG_PROC_FAILURE_BIT/MPIR_TAG_COLL_BIT` are  always placed into 31st - 29-th bits of user-provided tag (e.g.: [comm_create_group.c#L54](https://github.com/pmodels/mpich/blob/master/src/mpi/comm/comm_create_group.c#L54)), meanwhile device/netmod may provide less than 32 bits. That leads to utility-bits trimming from the tag on the device level and problems with tag matching on the reciever side:
`
Assertion failed in file ./src/mpi/coll/helper_fns.c at line 227: status->MPI_TAG == tag
`

### CODE REVIEW

**Comment by gcongiu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by gcongiu** [comment]
test:jenkins/ch3/ofi

**Comment by raffenet** [comment]
@gcongiu it looks like the izem and hwloc submodules were accidentally updated in this commit

**Comment by gcongiu** [comment]
@raffenet sorry about that, I will make sure this is fixed before asking for review

**Comment by gcongiu** [comment]
test:jenkins/ch3/all
test:jenkins/ch4/most

File: src/mpid/ch4/src/ch4_types.h [hunk 1]
```diff
@@ -296,7 +296,6 @@ typedef struct MPIDI_CH4_Global_t {
     int my_sigusr1_count;
 #endif
     OPA_int_t progress_count;
-
```
Comment (raffenet) [inline src/mpid/ch4/src/ch4_types.h]: Unrelated whitespace change.

File: src/mpid/ch3/channels/nemesis/netmod/ofi/ofi_tag_layout.h [hunk 2]
```diff
@@ -30,7 +30,7 @@
 #define MPID_PGID_SHIFT          (32)
 #define MPID_PSOURCE_SHIFT       (0)
 #define MPID_PORT_SHIFT          (0)
-#define MPID_TAG_BITS            (28)
+#define MPID_OFI_TAG_BITS        (28)
```
Comment (raffenet) [inline src/mpid/ch3/channels/nemesis/netmod/ofi/ofi_tag_layout.h]: For consistency, can you prefix all the constants here with `MPID_OFI_` in a patch before this one?

File: src/include/mpir_tags.h [hunk 3]
```diff
@@ -44,35 +44,32 @@
...
-#define MPIR_TAG_ERROR_BIT
-#endif
-
+#define MPIR_TAG_ERROR_BIT (1 << (MPID_TAG_BITS-1))
 /* This bitmask is used to differentiate between a process failure
  * (MPIX_ERR_PROC_FAILED) and any other kind of failure (MPI_ERR_OTHER). */
-#ifdef HAVE_TAG_ERROR_BITS
-#define MPIR_TAG_PROC_FAILURE_BIT (1 << 29)
+#define MPIR_TAG_PROC_FAILURE_BIT (1 << (MPID_TAG_BITS-2))
+#define MPIR_TAG_ERROR_BITS (2)
```
Comment (raffenet) [inline src/include/mpir_tags.h]: This change is related, but I'm leaning towards it being a separate commit prior to this commit. This change paves the way to making the tag space more dynamic.

Comment (gcongiu) [reply src/include/mpir_tags.h]: sorry I did not see this

**Comment by gcongiu** [comment]
test:jenkins/ch3/ofi

File: src/include/mpir_tags.h (hunk 3 above)
Comment (raffenet) [reply src/include/mpir_tags.h]: Just noting our discussion that this needs to be part of the commit that defines `MPID_TAG_BITS`, so my comment can be disregarded.

**Comment by gcongiu** [comment]
test:jenkins/ch3/ofi

**Comment by gcongiu** [comment]
@raffenet I made the changes to `ch3/ofi` and tests have passed

**Comment by raffenet** [comment]
OK, I am looking at this one more time, and I'm thinking that it would be more useful for the NM and SHM init hooks to return the number of tag bits they support, rather than a tag ub value. This way the upper layer can determine the tag_ub based on the number of error bits. I think this would come out a little cleaner. @gcongiu what do you think?

**Comment by gcongiu** [comment]
@raffenet I agree, having both `tag_ub` and `tag_bits` in `MPIR_Process` does not make too much sense.

**Comment by gcongiu** [comment]
@raffenet sounds good. I will do the changes

**Comment by gcongiu** [comment]
test:jenkins/ch3/ofi

**Comment by gcongiu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by gcongiu** [comment]
test:jenkins/ch3/most
test:jenkins/ch3/ofi
test:jenkins/ch4/most

**Comment by raffenet** [comment]
@gcongiu sorry to keep going around and around on this, but I don't think open coding the conversion from tag bits to tag upper bound is a good idea. It is too error prone. We should either add a macro for the conversion, or just continue to store the tag ub like before after calculating it once after tag bits is set.

**Comment by gcongiu** [comment]
@raffenet makes sense. Storing it would save some instruction so I guess this would be the best option?

**Comment by gcongiu** [comment]
@raffenet I have fixed the issue with `tag_ub`. Can you please have a look?

**Comment by gcongiu** [comment]
test:jenkins/ch3/ofi
test:jenkins/ch4/ofi

**Comment by gcongiu** [comment]
@raffenet this seems fine for merge. Maybe you want to have a final look at it just to make sure I haven't overlooked anything.

**Comment by raffenet** [comment]
test:jenkins/ch4/ucx

File: src/mpi/init/initthread.c [hunk 4]
```diff
@@ -431,6 +431,7 @@ int MPIR_Init_thread(int *argc, char ***argv, int required, int *provided)
     MPIR_Process.attrs.io = MPI_PROC_NULL;
     MPIR_Process.attrs.lastusedcode = MPI_ERR_LASTCODE;
     MPIR_Process.attrs.tag_ub = MPIR_TAG_USABLE_BITS;
+    MPIR_Process.attrs.tag_bits = MPID_TAG_BITS;
```
Comment (raffenet) [inline src/mpi/init/initthread.c]: I'm a little confused by this. So we set the tag_bits value to the device defined value `MPID_TAG_BITS`, but later on, the device defines `MPID_TAG_BITS` to be this global variable? There's got to be a clearer way of defining this...

File: src/mpid/ch3/include/mpidpre.h [hunk 5]
```diff
@@ -74,6 +74,7 @@ typedef int16_t MPIDI_Rank_t;
 #elif CH3_RANK_BITS == 32
 typedef int32_t MPIDI_Rank_t;
 #endif /* CH3_RANK_BITS */
+#define MPID_TAG_BITS (MPIR_Process.attrs.tag_bits)
```
Comment (raffenet) [inline src/mpid/ch3/include/mpidpre.h]: See here.

File: src/mpid/ch4/netmod/ucx/ucx_init.h [hunk 6]
```diff
@@ -113,7 +113,7 @@ static inline int MPIDI_NM_mpi_init_hook(int rank,
 #endif
 
     /* we reserve one bit to differentiate AMs from native messages */
-    *tag_ub = MPIDI_UCX_TAG_USABLE_BITS;
+    *tag_bits = MPID_TAG_BITS;
```
Comment (raffenet) [inline src/mpid/ch4/netmod/ucx/ucx_init.h]: I'm not understanding how this is reserving a bit. Obviously it seems to be working from the Jenkins results, but I don't quite get it.

Comment (gcongiu) [reply src/mpid/ch4/netmod/ucx/ucx_init.h]: This is just assigning the `tag_bits`, it is not reserving anything. I should have removed the comment. The restriction is on the `tag_ub` which, apparently, I forgot to fix. UCX netmod should overwrite `tag_ub` to a smaller value to reserve the `AM` bit. Probably it is not showing up in Jenkins because the test that uses the `MPI_TAG_UB` is not integrated yet (?)

Comment (raffenet) [reply src/mpid/ch4/netmod/ucx/ucx_init.h]: `tag_ub` is calculated as a function of the number of `tag_bits` available. The device should be reporting how many tag bits are supported. It is up to the upper layer to determine the `tab_ub`. Can you combine the large tag value test with this PR? I don't think we should merge this without it.

File: src/mpi/init/initthread.c (hunk 4 above)
Comment (gcongiu) [reply src/mpi/init/initthread.c]: If I am not mistaken this should be the default (i.e., 31). The netmod should overwrite these values later on. I think this is a mistake `MPIR_Process.attrs.tab_bits` is set by the device to default (see `src/mpid/ch4/src/ch4_init.c`). So both `MPIR_Process.attrs.tag_ub` and `MPIR_Process.attrs.tag_bits` should be assigned in `MPID_Init` to default and then overwritten, if necessary, by the netmod. This must be an error I did when reintroducing the `tag_ub` in MPIR_Process after removing it. Sorry for the confusion

Comment (raffenet) [reply src/mpi/init/initthread.c]: To be clear, I think it's perfectly fine to store `tag_ub` in the `MPIR_Process.attrs` object. We should set it once (based on the number of tag bits supported by the device - the number of reserved bits) and reuse it when needed.

**Comment by gcongiu** [comment]
test:jenkins/ch3/ofi
test:jenkins/ch4/ofi
test:jenkins/ch4/ucx

File: src/mpid/ch4/netmod/ucx/ucx_types.h [hunk 7]
```diff
@@ -23,8 +23,8 @@
 /* Active Message Stuff */
 #define MPIDI_UCX_NUM_AM_BUFFERS       (64)
 #define MPIDI_UCX_MAX_AM_EAGER_SZ      (16*1024)
-#define MPIDI_UCX_TAG_USABLE_BITS      (MPIR_TAG_USABLE_BITS >> 1)
-#define MPIDI_UCX_AM_TAG               (MPIDI_UCX_TAG_USABLE_BITS + 1)
+#define MPIDI_UCX_TAG_BITS             (MPIR_TAG_BITS_DEFAULT)
+#define MPIDI_UCX_AM_TAG               (MPIR_TAG_USABLE_BITS + 1)
```
Comment (gcongiu) [inline src/mpid/ch4/netmod/ucx/ucx_types.h]: @raffenet for consistency with `ch4/ofi` I have removed `MPIDI_UCX_TAG_USABLE_BITS`, which is now replaced by `MPIR_TAB_USABLE_BITS` and defined `MPDI_UCX_TAB_BITS`, to which `tag_bits` is set to in the UCX init function.

**Comment by gcongiu** [comment]
test:jenkins/ch4/ucx
test:jenkins/ch4/ofi

**Comment by gcongiu** [comment]
test:jenkins/ch4/ucx

**Comment by gcongiu** [comment]
test:jenkins/ch4/ucx

**Comment by gcongiu** [comment]
test:jenkins/ch4/ofi

**Comment by gcongiu** [comment]
test:jenkins/ch4/ofi
test:jenkins/ch4/ucx

File: src/mpid/ch4/netmod/ucx/ucx_types.h [hunk 8]
```diff
@@ -23,8 +23,12 @@
 /* Active Message Stuff */
 #define MPIDI_UCX_NUM_AM_BUFFERS       (64)
 #define MPIDI_UCX_MAX_AM_EAGER_SZ      (16*1024)
-#define MPIDI_UCX_TAG_USABLE_BITS      (MPIR_TAG_USABLE_BITS >> 1)
-#define MPIDI_UCX_AM_TAG               (MPIDI_UCX_TAG_USABLE_BITS + 1)
+#ifdef MPID_TAG_AUX_BITS
+#undef MPID_TAG_AUX_BITS
+#define MPID_TAG_AUX_BITS              (1)      /* we reserve one bit to differentiate AMs from native messages */
```
Comment (gcongiu) [inline src/mpid/ch4/netmod/ucx/ucx_types.h]: @raffenet sorry I haven't been able to clearly explain my point to you earlier. The only reasonable solution I could find for UCX AM tag bit (from my perspective) is to allow the netmod to reserve some auxiliary bits for dedicated functionalities. These are accounted for by `MPID_TAG_AUX_BITS`. Most netmods just use the default (`0`), UCX redefines it to `1`. In this way UCX netmod is still using `31` bits for the tag and not `30`. Since the position of the error and collective bits is now relative to the length of the tag this it cannot be changed to `30`. What changes is the `MPIR_TAG_USABLE_BITS`, which defines `tag_ub`. This is the only variable that has to account for the AM tag in UCX, or any other tag bit reserved by any other netmod. 
The alternative solution, from my perspective, would be to change the UCX AM code to remove the AM bit from the tag.

Comment (raffenet) [reply src/mpid/ch4/netmod/ucx/ucx_types.h]: What I don't understand is why `ch4/ofi` doesn't need to do this. It reserves 4 bits for dedicated functionality, doesn't it?

Comment (raffenet) [reply src/mpid/ch4/netmod/ucx/ucx_types.h]: Actually, no. I am mixing up the OFI AM header with the tag bits so that's not the issue.

Comment (raffenet) [reply src/mpid/ch4/netmod/ucx/ucx_types.h]: OK, I think I understand now that OFI netmod defines these protocol bits outside of the user tag space, whereas UCX netmod does not. I think your suggestion to move the AM bit outside of the tag is the right thing to do for simplicity.

Comment (gcongiu) [reply src/mpid/ch4/netmod/ucx/ucx_types.h]: I am not sure how much changes are required for this. Should this be a separate PR?

Comment (raffenet) [reply src/mpid/ch4/netmod/ucx/ucx_types.h]: I think it can be part of this PR. Moving the AM bit outside of the user tag should not be a big burden. It can be done prior to the alignment commits.

**Comment by gcongiu** [comment]
test:jenkins/ch4/ucx

File: src/mpid/ch4/netmod/ucx/ucx_types.h [hunk 9]
```diff
@@ -53,11 +51,12 @@ extern ucp_generic_dt_ops_t MPIDI_UCX_datatype_ops;
...
 #define MPIDI_UCX_CONTEXT_TAG_BITS 16
 #define MPIDI_UCX_CONTEXT_RANK_BITS 16
-#define UCX_TAG_BITS 32
+#define UCX_TAG_BITS 30
 
-#define MPIDI_UCX_TAG_MASK      (0x00000000FFFFFFFFULL)
-#define MPIDI_UCX_SOURCE_MASK   (0x0000FFFF00000000ULL)
-#define MPIDI_UCX_TAG_SHIFT     (32)
+#define MPIDI_UCX_TAG_MASK      (0x000000007FFFFFFFULL)
+#define MPIDI_UCX_SOURCE_MASK   (0x00007FFF80000000ULL)
```
Comment (gcongiu) [inline src/mpid/ch4/netmod/ucx/ucx_types.h]: @raffenet I think this might work. I have moved the AM bit from the tag to the MSB in the matching bits, thus I had to change the bit masks.

**Comment by gcongiu** [comment]
test:jenkins/ch4/ucx
test:jenkins/ch4/ofi

File: src/mpid/ch4/netmod/ucx/ucx_types.h (hunk 9 above)
Comment (raffenet) [reply src/mpid/ch4/netmod/ucx/ucx_types.h]: Can you update lines 48-50 with the new layout?

Comment (gcongiu) [reply src/mpid/ch4/netmod/ucx/ucx_types.h]: Done

Comment (raffenet) [reply src/mpid/ch4/netmod/ucx/ucx_types.h]: OK, one last dumb question. The layout indicates 31 tag bits, but `UCX_TAG_BITS` is 30?

Comment (gcongiu) [reply src/mpid/ch4/netmod/ucx/ucx_types.h]: The number of effective tag bits is 30. The 31st is reserved for some reason I suppose. In fact, the error bit is always the second MSB of the tag field not the first. Probably the way I am assigning the number of tag bits is confusing. Earlier on the `UCX_TAG_BITS` constant was assigned but unused and the number of actual tag bits was 31 + the AM bit.

Comment (gcongiu) [reply src/mpid/ch4/netmod/ucx/ucx_types.h]: Actually forget about it, you are right is 31. Let me fix it

**Comment by gcongiu** [comment]
test:jenkins/ch4/ucx

File: src/mpid/ch4/netmod/ucx/ucx_impl.h [hunk 10]
```diff
@@ -33,7 +33,7 @@ static inline uint64_t MPIDI_UCX_init_tag(MPIR_Context_id_t contextid, int sourc
     ucp_tag = (ucp_tag << MPIDI_UCX_SOURCE_SHIFT);
     ucp_tag |= source;
     ucp_tag = (ucp_tag << MPIDI_UCX_TAG_SHIFT);
-    ucp_tag |= (MPIDI_UCX_TAG_MASK & tag);
+    ucp_tag |= tag;
```
Comment (raffenet) [inline src/mpid/ch4/netmod/ucx/ucx_impl.h]: Won't this overwrite some of the source bits? Can we add a Min's tag_ub test to this PR to verify everything is working correctly?

Comment (raffenet) [reply src/mpid/ch4/netmod/ucx/ucx_impl.h]: Come to think of it, the tag_ub test won't necessarily fail here. I still think we should test it, though.

But I think the issue is clear. The tag mask now covers 31 bits after modifying above. Removing its use here means we will set the lower 32 bits in `ucp_tag`, one more than we should.

**Comment by gcongiu** [comment]
test:jenkins/ch4/ucx

**Comment by raffenet** [comment]
Superceded by #3349.
